---
title: "Bioconductor Exercises"
author: "MRC CSC Bioinformatics Core Team"
date: "19 July 2016"
output: html_document
---

----
```{r echo=FALSE, eval=TRUE, warning=FALSE}

```


# Exercises

1. Check what kind of annotation data can be extracted using `org.Hs.eg.db` package
2. Print few gene names from `org.Hs.eg.db`
3. Retrieve gene name, chromosome and Ensembl gene identifiers for “HEBP2” and “PRND” from `org.Hs.eg.db`
4. Retrieve gene symbol, gene name and gene alias for genes in chromosome 2 for human using `org.Hs.eg.db`
5. Print non-redundant list of chromosomes from `org.Mm.eg.db`
6. How many annotation datasets available in Ensembl Biomart?
7. Retrieve genomic coordinates for human genes from Ensembl biomart and build a GRanges object.
    + Subset the above GRanges object to include only protein coding genes
    + Subset the GRanges object again with genes in main chromsomes (1-22,X,Y)
    + Create another GRanges object with genes in chr1:1544000-2371000
8. Retrieve 200 bp upstream promoter sequences for the given gene symbols `AQP1, ASNSP2, KPNA2, FRMD4A, NSUN5, VAC14` from Ensembl human biomart.
    + Tips: 
    + Read documentation for `getSequence`
    + Use `type="hgnc_symbol"` and `seqType="coding_gene_flank"`
9. Retrieve the transcript coordinates for genes as GRangesList from `TxDb.Hsapiens.UCSC.hg19.knownGene` (install it from Bioconductor if required)
10. Retrieve exon coordiantes for genes from `TxDb.Hsapiens.UCSC.hg19.knownGene`

    
8. Import human hg19 gene coordiantes from "hg19Genes.txt and create a GRanges of Transcription start sites (1 bp range).
    + Tip: How to identify TSS for genes in forward/reverse strand?
9. Create a GRanges object of human promoters with TSS ± 2000bp (using the GRanges object created in Q8).
    + Tip: Read the documentation for `promoters` function.
11. Import ELF1 binding sites in K562 cell from Encode (ELF1_K562.bed) and create GRanges object. 
    + Tips: Import the ELF1 binding sites using `import.bed()` function from `rtracklayer` package
    + Find ELF1 binding sites overlap with promoters (TSS ± 1kb) using `findOverlaps` and `subsetByOverlaps` (Remember BED format uses 0-based coordinates)
    





## **Additional Exercise**

* Download the following BAM and index files (.bai) (ENCODE data - ChIP-Seq of CTCF in Ag04449 human fibroblast cells).
  + http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwTfbs/wgEncodeUwTfbsAg04449CtcfStdAlnRep1.bam
  + http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwTfbs/wgEncodeUwTfbsAg04449CtcfStdAlnRep1.bam.bai
* Use `readGAlignments` to read the bam. Construct an `ScanBamParam` object that accepts only aligned reads, passing quality control and not duplicates.
* Compute genome wide coverage using `coverage` function
* Export coverage as bigWig using `export.bw()` from rtracklayer package and visualise it using IGV
* Compute number of reads overlapping with hg19 promoters (TSS ± 1kb) and export the results as text file.


<br><br>

## Solutions

#### 1. Check what kind of annotation data can be extracted using `org.Hs.eg.db` package
```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE}
library("org.Hs.eg.db")
columns(org.Hs.eg.db)
```

#### 2. Print few gene names from `org.Hs.eg.db`
```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE}
head(keys(org.Hs.eg.db,keytype="GENENAME"))
```

#### 3. Retrieve gene name, chromosome and Ensembl gene identifiers for “HEBP2” and “PRND” from `org.Hs.eg.db`

```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE, warning=FALSE}
select(org.Hs.eg.db,keys=c("HEBP2","PRND"),keytype="SYMBOL",
       columns=c("GENENAME","CHR","ENSEMBL"))
```

#### 4. Retrieve gene symbol, gene name and gene alias for genes in chromosome 2 for human using `org.Hs.eg.db`
```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE, warning=FALSE}
chr2Genes <- select(org.Hs.eg.db,keys="2",keytype="CHR",
                    columns=c("SYMBOL","GENENAME","ALIAS"))
head(chr2Genes)
```

#### 5. Print non-redundant list of chromosomes from `org.Mm.eg.db`
```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE, warning=FALSE}
library("org.Mm.eg.db")
unique(keys(org.Mm.eg.db,keytype="CHR"))
```

#### 6. How many annotation datasets available in Ensembl Biomart?
```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE, warning=FALSE}
library("biomaRt")
ensembl <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", host = "sep2015.archive.ensembl.org") # select ensembl
ens_datasets <- listDatasets(ensembl) # list datasets 
head(ens_datasets)
dim(ens_datasets)
```

#### 7. Retrieve genomic coordinates for human genes from Ensembl biomart and build a GRanges object.
  + Subset the above GRanges object to include only protein coding genes
  + Subset the GRanges object again with genes in main chromsomes (1-22,X,Y)
  + Create another GRanges object with genes in chr1:1544000-2371000
      
      + Tips:
      + You can select main chromosomes and "protein coding" genes by using appropriate filter and value.
      + Search for "biotype" in available filters using `grep()`
      + Run `filterOptions("biotype",selectedmart)` to see the accepted values for "biotype" filter
      + When multiple filters specified, "values" argument should be a list of vectors; each vector corresponds to each specified filter.
      + Annotation fields to retrieve: "chromosome_name", "start_position", "end_position","ensembl_gene_id", "strand", "external_gene_name"
      + Before creating GRanges object, add "chr" prefix to chromosome using `paste` function, Ex: change 1 to chr1 (required for next task)

```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE, warning=FALSE}
ens_human <- useDataset("hsapiens_gene_ensembl",mart=ensembl) # select human dataset
ens_human_Attr <- listAttributes(ens_human) # list available annotation
ens_human_filters <- listFilters(ens_human) # list available filters
availFilters <- filterOptions("biotype",ens_human) # Displays accepted values for "biotype"

hg19Gene <- getBM(
          attributes = c("chromosome_name","start_position","end_position",
                         "ensembl_gene_id","strand","external_gene_name"), 
          filter=c("chromosome_name","biotype"),
          values=list(c(1:22,"X","Y"),"protein_coding"), mart=ens_human)
head(hg19Gene)
```

Now create GRanges object using the above data frame. 
```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE}
library(GenomicRanges)

# add 'chr' prefix to chromosome name
hg19Gene$chromosome_name <- paste("chr",hg19Gene$chromosome_name,sep="")

hg19Gene.GR <- GRanges(seqnames=hg19Gene$chromosome_name,
                       ranges=IRanges(start=hg19Gene$start_position,end=hg19Gene$end_position),
                       strand=ifelse(hg19Gene$strand==1,"+","-"),
                       EnsemblID=hg19Gene$ensembl_gene_id,
                       Symbol=hg19Gene$external_gene_name)
hg19Gene.GR
```

Filter the above GRanges object for genes in chr1:1544000-2371000
```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE}
chr1genes <- hg19Gene.GR[seqnames(hg19Gene.GR)=="chr1" & 
                         start(hg19Gene.GR) > 1544000 & 
                           end(hg19Gene.GR) < 2371000]
head(chr1genes)

# alternate solution
chr1genes <- subset(hg19Gene.GR,start>1544000 & end<2371000 & seqnames=="chr1")
```


#### 8. Create a GRanges of Transcription start sites (1 bp range) for the GRanges object created in Q7.
    + Tip: How to identify TSS for genes in forward/reverse strand?

```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE}
hg19Gene$TSS <- ifelse(hg19Gene$strand==1,hg19Gene$start_position,hg19Gene$end_position)

hg19TSS <- GRanges(seqnames=hg19Gene$chromosome_name,
                       ranges=IRanges(start=hg19Gene$TSS,end=hg19Gene$TSS),
                       strand=ifelse(hg19Gene$strand==1,"+","-"),
                       EnsemblID=hg19Gene$ensembl_gene_id,
                       Symbol=hg19Gene$external_gene_name)
hg19TSS
```


#### 9. Create a GRanges object of human promoters with TSS ± 2000bp (using the GRanges object created in Q7).
    + Tip: Read the documentation for `promoters` function.
```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE}
hg19Promoters <- promoters(hg19Gene.GR,upstream=2000,downstream=2000)
hg19Promoters
```

#### 10. Retrieve 200 bp upstream promoter sequences for the given gene symbols `AQP1, ASNSP2, KPNA2, FRMD4A, NSUN5, VAC14` from Ensembl human biomart.
    + Tips: 
    + Read documentation for `getSequence`
    + Use `type="hgnc_symbol"` and `seqType="coding_gene_flank"`

```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE, warning=FALSE}
symbols <- c("AQP1", "ASNSP2", "KPNA2", "FRMD4A", "NSUN5", "VAC14")
ensembl <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", host = "jul2015.archive.ensembl.org", dataset="hsapiens_gene_ensembl") # select ensembl

seq <- getSequence(id=symbols, type="hgnc_symbol",
                  seqType="coding_gene_flank", upstream=200, mart = ensembl)
```


#### 11. Import ELF1 binding sites in K562 cell from Encode (ELF1_K562.bed) and create GRanges object. 
    + Tips: Import the ELF1 binding sites using `import.bed()` function from `rtracklayer` package
    + Find ELF1 binding sites overlap with promoters (TSS ± 1kb) using `findOverlaps` and `subsetByOverlaps` (Remember BED format uses 0-based coordinates)

```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE}
library("rtracklayer")
ELF1 <- read.table("ELF1_K562.bed",sep="\t",header=F)
ELF1GR <- GRanges(seqnames=ELF1$V1, IRanges(start=ELF1$V2+1,end=ELF1$V3))
ELF1GR_A <- import.bed("ELF1_K562.bed")


# ELF1 binding sites overlap with promoters using `findOverlaps`
hg19Promoters <- promoters(hg19Gene.GR,upstream=1000,downstream=1000)
ELF1GR <- reduce(ELF1GR) # merging overlapping peaks
ELF1overlap <- findOverlaps(ELF1GR,hg19Promoters,ignore.strand=T)
ELF1overlap.m <- as.matrix(ELF1overlap)
ELF1_promoters <- ELF1GR[ELF1overlap.m[,"queryHits"],]
ELF1_promoters

# ELF1 binding sites overlap with promoters using `subsetByOverlaps`
ELF1_promoters1 <- subsetByOverlaps(ELF1GR,hg19Promoters)
ELF1_promoters1
```

<b>Note the differences in the outputs!</b>

#### 12. Retrieve the transcript coordinates for genes as GRangesList from `TxDb.Hsapiens.UCSC.hg19.knownGene` (install it from Bioconductor if required)
```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE}
# source("http://bioconductor.org/biocLite.R")
# biocLite("TxDb.Hsapiens.UCSC.hg19.knownGene")
library("TxDb.Hsapiens.UCSC.hg19.knownGene")
hg19txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
TranscrtipsByGene <- transcriptsBy(hg19txdb,by="gene") # inspect the output
```

#### 13. Retrive exon coordiantes for genes from `TxDb.Hsapiens.UCSC.hg19.knownGene`
```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE}
ExonsByGene <- exonsBy(hg19txdb,by="gene") # inspect the output
```





## Additional Exercise

* Download the following BAM and index files (*.bai) (ENCODE data - ChIP-Seq of CTCF in Ag04449 human fibroblast cells)
    - http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwTfbs/wgEncodeUwTfbsAg04449CtcfStdAlnRep1.bam
    - http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwTfbs/wgEncodeUwTfbsAg04449CtcfStdAlnRep1.bam.bai
* Use `readGAlignments` to read the bam. Construct an `ScanBamParam` object that accepts only aligned reads, passing quality control and not duplicates.
* Compute genome wide coverage using `coverage` function
* Export coverage as bigWig using `export.bw()` from rtracklayer package and visualise it using IGV
* Compute number of reads overlapping with hg19 promoters (TSS ± 1kb) and export the results as text file.

```{r, echo=TRUE, eval=FALSE, message=FALSE, collapse=TRUE}
library("GenomicAlignments")
library("rtracklayer")
bamFile <- "wgEncodeUwTfbsAg04449CtcfStdAlnRep1.bam"
flag <- scanBamFlag()
param <- ScanBamParam(
    flag=scanBamFlag(isUnmappedQuery=FALSE, isDuplicate=FALSE, isNotPassingQualityControls=FALSE)
    )

# Read the BAM
CTCF <- readGAlignments(bamFile,param=param)

# Generate the genomic coverage and inspect the output
CTCFCov <- coverage(CTCF)

# Export as bigWig
export.bw(CTCFCov, "Ag04449_CTCF.bw")

# Reads overlapping with promoters
CTCFCounts <- countOverlaps(hg19Promoters,CTCF)

# Add CTCF counts as elementMetadata to hg19Promoters object
mcols(hg19Promoters)$CTCF <- CTCFCounts

# Export the results as text file
hg19Promoters.df <- as.data.frame(hg19Promoters)
write.table(hg19Promoters.df,"hg19Promoters_CTCF.txt",sep="\t",row.names=F)

```


Session Info
----------------------------------------------------------------------
```{r, echo=TRUE, eval=TRUE, message=FALSE, collapse=TRUE}
sessionInfo()
```