---
title: "Bioconductor Exercises Part 2"
author: "MRC LMS Bioinformatics Core"
date: "19 September 2019"
output: html_document
---

----
```{r echo=FALSE, eval=TRUE, warning=FALSE}

```


# Exercises

* Download the following BAM (aligned to hg19) and index files (.bai) (ENCODE data - ChIP-Seq of CTCF in Ag04449 human fibroblast cells).
  + http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwTfbs/wgEncodeUwTfbsAg04449CtcfStdAlnRep1.bam
  + http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeUwTfbs/wgEncodeUwTfbsAg04449CtcfStdAlnRep1.bam.bai
* Use `readGAlignments` to read the bam. Construct an `ScanBamParam` object that accepts only aligned reads, passing quality control and not duplicates.
* Compute genome wide coverage using `coverage` function
* Export coverage as bigWig using `export.bw()` from rtracklayer package and visualise it using IGV (https://software.broadinstitute.org/software/igv/download)
* Compute number of reads overlapping with hg19 promoters (TSS Â± 1kb) and export the results as csv file. [hint: use "./data/hg19Genes.txt" to create the hg19 promoters]


    
